### PR TITLE
Refuse to load fields from _source when using the `fields` option and support wildcards. 

### LINKED ISSUE #10783: Support * wildcard to retrieve stored fields
Currently you can retrieve stored fields of a search hit using the [fields](http://www.elastic.co/guide/en/elasticsearch/reference/current/search-request-fields.html) option.
However you cannot use wildcards.

Suppose I have document with several title fields such as "book_title", "chapter_title", "paragraph_title", "section_title" and I want to have all those title fields returned from a query.

It would be nice to have something like this:

```
{
    "fields" : [ "*_title" ],
    "query" : {
        "term" : { "category" : "quote" }
    }
}
```

No need to specify all the fields and when new title fields are added they are automatically returned.
Wildcard * is already supported when specifying the fields in a multi_match query or when specifying the fields to highlight for example.

A current workaround is using [source-filtering](http://www.elastic.co/guide/en/elasticsearch/reference/current/search-request-source-filtering.html), which does allow the * wildcard. 

```
{
    "_source": "*_title" ,
    "query" : {
        "term" : { "category" : "quote" }
    }
}
```

### LINKED ISSUE #14489: Refuse to load fields from _source when using the `fields` option
Spin-off from #14474. The `fields` option is intended to load stored fields. However it will try to load these fields from the source if they are not stored. This causes confusion because then eg. wildcards don't work.

We should deprecate the ability to load fields from _source and remove it in 3.0.

### PR DESCRIPTION
**Pull request by jimczi**
Fixes #14489  
  Do not to load fields from _source when using the `fields` option.
  Non stored (non existing) fields are ignored by the fields visitor when using the `fields` option.

Fixes #10783
  Support \* wildcard to retrieve stored fields when using the `fields` option.
  Supported pattern styles are "xxx_", "_xxx", "_xxx_" and "xxx*yyy".

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/index/fieldvisitor/WildcardFieldsVisitor.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+import java.util.List;
+import java.util.Set;
+
+
+/**
+ * A field visitor that allows to load a selection of the stored
+ * fields by names or by patterns.
+ * Supported pattern styles "xxx*", "*xxx", "*xxx*" and "xxx*yyy".
+ */
+public class WildcardFieldsVisitor extends CustomFieldsVisitor {
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/index/fieldvisitor/WildcardFieldsVisitor.java]: Should we fold the functionality directly into CustomFieldsVisitor?

File: core/src/test/java/org/elasticsearch/search/aggregations/metrics/TopHitsIT.java [hunk 2]
```diff
@@ -532,8 +532,8 @@ public void testFetchFeatures() {
                                         topHits("hits").setSize(1)
                                             .highlighter(new HighlightBuilder().field("text"))
                                             .setExplain(true)
+                                                .addField("text")
```
Comment (jpountz) [inline core/src/test/java/org/elasticsearch/search/aggregations/metrics/TopHitsIT.java]: can you align indentation with the above/below lines?

File: core/src/main/java/org/elasticsearch/index/fieldvisitor/CustomFieldsVisitor.java [hunk 3]
```diff
@@ -43,6 +43,6 @@ public Status needsField(FieldInfo fieldInfo) throws IOException {
             return Status.YES;
         }
 
-        return fields.contains(fieldInfo.name) ? Status.YES : Status.NO;
+        return fields == null ? Status.NO : fields.contains(fieldInfo.name) ? Status.YES : Status.NO;
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/index/fieldvisitor/CustomFieldsVisitor.java]: Why do we need the null check? Should we requires that `fields` is not null instead and fix callers to pass `Collections.emptySet()` when they don't want to load any field?

File: core/src/main/java/org/elasticsearch/index/fieldvisitor/WildcardFieldsVisitor.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    private List<String> patterns;
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/index/fieldvisitor/WildcardFieldsVisitor.java]: can you make it final?

**Comment by jpountz** [comment]
Thanks Jim, I'm happily surprised that there were not too many tests using the `fields` option to get fields from the `_source`! Could you add a note to `docs/reference/migration/migrate_3_0.asciidoc` to let users know that in 3.0 the `fields` option won't be able to load fields from `_source` anymore?

File: rest-api-spec/src/main/resources/rest-api-spec/test/search/10_source_filtering.yaml [hunk 5]
```diff
@@ -87,7 +86,7 @@
           body:
             fields: [ include.field2, _source ]
             query: { match_all: {} }
-  - match: { hits.hits.0.fields: { include.field2 : [v2] }}
+  - match: { hits.hits.0._source.include.field2: v2 }
   - is_true:  hits.hits.0._source
```
Comment (jpountz) [inline rest-api-spec/src/main/resources/rest-api-spec/test/search/10_source_filtering.yaml]: The point of the two above tests seems to be to test bw compat, I think we can remove them?

**Comment by jpountz** [comment]
Thanks @jimferenczi this looks good to me. I just left a comment about two REST tests that I think we should remove. @bleskes would you mind having a quick look at this PR before it gets merged?

**Comment by bleskes** [comment]
This looks good to me. Thanks @jimferenczi . My only question is whether we should resolve wild cards in the field visitor. Feels more natural for me to resolve it based on the mappings and get a concrete list to the visitor. I think we have the support we need in the MapperService (or will be easy to build). I'm worried that a wild card on the lucene level may expose system/hidden fields users didn't mean.. I don't think it's a problem in practice though. @jpountz wdyt?

File: core/src/main/java/org/elasticsearch/search/fetch/FetchPhase.java [hunk 6]
```diff
@@ -123,32 +111,28 @@ public void execute(SearchContext context) {
...
-                } else {
-                    if (extractFieldNames == null) {
-                        extractFieldNames = new ArrayList<>();
-                    }
-                    extractFieldNames.add(fieldName);
+                    fieldNames.add(fieldName);
                 }
             }
-            if (loadAllStored) {
-                fieldsVisitor = new AllFieldsVisitor(); // load everything, including _source
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/search/fetch/FetchPhase.java]: is AllFieldsVisitor still used with your change? If not can you remove it?

**Comment by jpountz** [comment]
@bleskes Good question. We already support fields=* in master (see mentions of `loadAllStoredFields` in the code) and I'm not aware of problems, so I *think* we should be fine.

File: core/src/main/java/org/elasticsearch/search/fetch/FetchPhase.java (hunk 6 above)
Comment (jimferenczi) [inline core/src/main/java/org/elasticsearch/search/fetch/FetchPhase.java]: no it's not. I'll remove it.

**Comment by jpountz** [comment]
@jimferenczi Let's merge this one?

**Comment by clintongormley** [comment]
@jimferenczi please add version labels for each branch that you merged this PR into

**Pull request merged**